Disallow double signs and spaces between a sign and number in my_atof3

t-a-k opened the pull request:
`Perl_my_atof3` used to pass a substring after the first (optional) sign to `S_strtod`, which causes wrong numifications for strings like `"-+3"` or `"+ 0x123"` (for the latter case, while `Perl_my_atof3 already` had the code to block `"0x"` prefixes, this string will slip through due to the space character in it).

Will fix #18584.

t-a-k committed:
Perl_my_atof3: disallow double signs and spaces between a sign and number

Perl_my_atof3 used to pass a substring after the first (optional) sign
to (S_)strtod, which causes wrong numifications for strings like "-+3"
or "+ 0x123" (for the latter case, while Perl_my_atof3 already had
the code to block "0x" prefixes, this string will slip through due to
the space character in it).

For GH #18584.

hvds commented on numeric.c:
```diff
@@ -1675,14 +1675,18 @@ Perl_my_atof3(pTHX_ const char* orig, NV* value, const STRLEN len)
             return (char *)s+1;
         }
 
+        /* strtod will parse a sign (and skip leading whitespaces) by itself,
+         * so rewind s to the beginning of the string. */
```
I think it is important to point out is the reason we _should_ let it parse them itself, ie that we'd get an undesirable result if we pointed it at some apparently-leading whitespace _after_ the sign; maybe something like this:
```
        /* We do not want strtod to parse whitespace after the sign, since
         * that would give backward-incompatible results. So we rewind and 
         * let strtod handle the whitespace and sign character itself. */
```

t-a-k replied:
Thank you, but I've forgotten that `"+-inf"` is still wrongly numified, so I will fix it too.

t-a-k committed:
Perl_my_atof3: disallow double signs for Inf/NaN

Perl_my_atof3 used to call S_my_atof_infnan after parsing sign character,
but this led Inf/NaN with double signs (e.g. "+-Inf") to be wrongly accepted
because S_my_atof_infnan will also parse sign itself.

Also improved comment per suggestion from @hvds.

t-a-k committed:
t/op/numify.t: More tests to check if invalid Inf/NaNs are not numified
